File: src/services/notificationService.js
// src/services/notificationService.js
// Implementa a tabela "Regras de notificação" (secção 6 da especificação).
// Cada evento de negócio chama uma função aqui — mantém a lógica de "quem é
// avisado, por que canal" num único sítio.

const prisma = require('../config/database');
const logger = require('../config/logger');

// Providers de email/in-app são plugáveis. No MVP, o EMAIL_PROVIDER=console
// apenas regista no log — trocar por um provider real (SMTP/SES/SendGrid)
// sem tocar no resto do código.
const config = require('../config/env');

async function dispatchEmail(to, subject, body) {
  if (config.email.provider === 'console' || !to) {
    logger.info('Email (modo console)', { to, subject, body });
    return;
  }
  // TODO: integrar provider SMTP real usando config.email.smtp
  logger.info('Email enviado', { to, subject });
}

async function notifyUser({ userId, type, title, message, channel = 'IN_APP', relatedEntityType, relatedEntityId, emailTo }) {
  const notification = await prisma.notification.create({
    data: {
      userId,
      type,
      channel,
      title,
      message,
      relatedEntityType,
      relatedEntityId,
    },
  });

  if (channel === 'IN_APP_EMAIL' && emailTo) {
    await dispatchEmail(emailTo, title, message);
  }

  return notification;
}

async function notifyUsersByRole({ companyId, roles, type, title, message, channel = 'IN_APP', relatedEntityType, relatedEntityId }) {
  const users = await prisma.user.findMany({
    where: { companyId, role: { in: roles }, active: true },
  });

  return Promise.all(
    users.map((u) =>
      notifyUser({
        userId: u.id,
        type,
        title,
        message,
        channel,
        relatedEntityType,
        relatedEntityId,
        emailTo: u.email,
      })
    )
  );
}

async function notifyCompanyContact({ companyId, type, title, message }) {
  const company = await prisma.company.findUnique({ where: { id: companyId } });
  if (!company) return null;

  const notification = await prisma.notification.create({
    data: { companyId, type, channel: 'EMAIL', title, message },
  });
  await dispatchEmail(company.contactEmail, title, message);
  return notification;
}

// --- Eventos de negócio (mapeados 1:1 com a secção 6 da especificação) -----

const events = {
  poAguardaAprovacao: (po) =>
    notifyUsersByRole({
      companyId: po.buyerCompanyId,
      roles: ['COMPANY_ADMIN'],
      type: 'PO_AGUARDA_APROVACAO',
      title: 'Nova PO aguarda aprovação',
      message: `A ordem de compra ${po.reference} aguarda a sua aprovação.`,
      channel: 'IN_APP',
      relatedEntityType: 'PurchaseOrder',
      relatedEntityId: po.id,
    }),

  poAprovadaOuRejeitada: (po) =>
    notifyUser({
      userId: po.createdById,
      type: po.status === 'APROVADA' ? 'PO_APROVADA' : 'PO_REJEITADA',
      title: po.status === 'APROVADA' ? 'PO aprovada' : 'PO rejeitada',
      message: `A ordem de compra ${po.reference} foi ${po.status === 'APROVADA' ? 'aprovada' : 'rejeitada'}.`,
      channel: 'IN_APP',
      relatedEntityType: 'PurchaseOrder',
      relatedEntityId: po.id,
    }),

  poRecebidaPeloFornecedor: (po) =>
    notifyUsersByRole({
      companyId: po.supplierCompanyId,
      roles: ['FORNECEDOR', 'COMPANY_ADMIN'],
      type: 'PO_RECEBIDA_FORNECEDOR',
      title: 'Nova ordem de compra recebida',
      message: `Recebeu a ordem de compra ${po.reference}. Reveja e aceite ou recuse.`,
      channel: 'IN_APP_EMAIL',
      relatedEntityType: 'PurchaseOrder',
      relatedEntityId: po.id,
    }),

  faturaGerada: (invoice, po) =>
    notifyUsersByRole({
      companyId: po.buyerCompanyId,
      roles: ['FINANCEIRO'],
      type: 'FATURA_GERADA',
      title: 'Fatura pendente de pagamento',
      message: `A fatura ${invoice.reference} (PO ${po.reference}) foi gerada. Prazo de pagamento: ${po.paymentDueAt?.toISOString().slice(0, 10)}.`,
      channel: 'IN_APP_EMAIL',
      relatedEntityType: 'Invoice',
      relatedEntityId: invoice.id,
    }),

  pagamentoProcessado: (payment, po) =>
    notifyUsersByRole({
      companyId: po.supplierCompanyId,
      roles: ['FORNECEDOR', 'COMPANY_ADMIN'],
      type: 'PAGAMENTO_PROCESSADO',
      title: 'Pagamento recebido',
      message: `O pagamento da PO ${po.reference} foi processado. Pode iniciar a execução/entrega.`,
      channel: 'IN_APP_EMAIL',
      relatedEntityType: 'Payment',
      relatedEntityId: payment.id,
    }),

  entregaDespachada: (po) =>
    notifyUsersByRole({
      companyId: po.buyerCompanyId,
      roles: ['COMPRADOR'],
      type: 'ENTREGA_DESPACHADA',
      title: 'Entrega despachada',
      message: `A entrega da PO ${po.reference} foi despachada.`,
      channel: 'IN_APP',
      relatedEntityType: 'PurchaseOrder',
      relatedEntityId: po.id,
    }),

  rececaoComDivergencia: (po) =>
    notifyUsersByRole({
      companyId: null,
      roles: ['ADMIN_SISTEMA'],
      type: 'RECECAO_COM_DIVERGENCIA',
      title: 'Receção com divergência reportada',
      message: `A PO ${po.reference} foi recebida com divergência ("${po.receptionStatus}"). Caso a acompanhar fora da plataforma (sinistro).`,
      channel: 'IN_APP_EMAIL',
      relatedEntityType: 'PurchaseOrder',
      relatedEntityId: po.id,
    }),

  apoliceSubmetidaOuAprovada: (companyId, policyLabel) =>
    notifyUsersByRole({
      companyId,
      roles: ['COMPANY_ADMIN', 'FINANCEIRO'],
      type: 'APOLICE_SUBMETIDA_APROVADA',
      title: 'Apólice atualizada',
      message: `A apólice ${policyLabel} foi submetida/aprovada.`,
      channel: 'IN_APP_EMAIL',
    }),

  apoliceAExpirar: (companyId, policyLabel, validUntil) =>
    notifyUsersByRole({
      companyId,
      roles: ['COMPANY_ADMIN', 'FINANCEIRO'],
      type: 'APOLICE_A_EXPIRAR',
      title: 'Apólice a expirar em breve',
      message: `A apólice ${policyLabel} expira em ${validUntil.toISOString().slice(0, 10)}. Providencie a renovação.`,
      channel: 'IN_APP_EMAIL',
    }),

  cadastroEmpresaDecidido: (company) =>
    notifyCompanyContact({
      companyId: company.id,
      type: company.status === 'APROVADA' ? 'CADASTRO_EMPRESA_APROVADO' : 'CADASTRO_EMPRESA_REJEITADO',
      title: company.status === 'APROVADA' ? 'Cadastro aprovado' : 'Cadastro rejeitado',
      message:
        company.status === 'APROVADA'
          ? `O cadastro da empresa ${company.name} foi aprovado. Já pode transacionar na KIXIMA.`
          : `O cadastro da empresa ${company.name} foi rejeitado.`,
    }),
};

module.exports = { notifyUser, notifyUsersByRole, notifyCompanyContact, events };

File: backend-java/src/main/java/ao/kixima/notification/NotificationService.java
package ao.kixima.notification;

import ao.kixima.company.Company;
import ao.kixima.company.CompanyRepository;
import ao.kixima.invoice.Invoice;
import ao.kixima.po.PoStatus;
import ao.kixima.po.PurchaseOrder;
import ao.kixima.security.PersonaRole;
import ao.kixima.supplierdev.SupplierDevRequest;
import ao.kixima.user.User;
import ao.kixima.user.UserRepository;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.UUID;

/**
 * Espelha backend/src/services/notificationService.js — implementa a
 * tabela "Regras de notificação" (secção 6 da especificação). Cada evento
 * de negócio chama um método aqui — mantém "quem é avisado, por que
 * canal" num único sítio.
 *
 * DIVERGÊNCIA DELIBERADA face ao Node: lá, os eventos são chamados DEPOIS
 * do {@code prisma.$transaction(...)} fechar (o negócio já está comitado
 * antes de notificar). Aqui, PoService chama estes métodos DENTRO da
 * própria transacção Spring — porque {@link EmailDispatchService#dispatch}
 * nunca lança (falha de envio fica só no log, mesmo princípio do Node) e a
 * escrita da {@link Notification} é uma simples INSERT na mesma base, sem
 * chamada de rede a um terceiro (ao contrário da submissão AGT, essa sim
 * feita depois do commit — ver PoService.acceptPurchaseOrder). Isto torna
 * "PO aprovada" + "notificação criada" atómicos, o que é estritamente mais
 * seguro; a única divergência real é o caso raro de a própria escrita da
 * notificação falhar por erro de base de dados, que aqui reverteria também
 * a operação de negócio, e no Node não.
 *
 * NÃO PORTADO NESTE MARCO (M5):
 * <ul>
 *   <li>{@code realtimeService.emitToUser} — push imediato via Socket.IO/
 *   STOMP; a notificação fica sempre gravada (fonte da verdade), só falta
 *   o "empurrão" em tempo real — entra no M6, junto com o resto de
 *   tempo real (decisão já tomada no plano).</li>
 *   <li>{@code i18n/emails.js} — tradução do assunto/corpo do EMAIL para o
 *   idioma do destinatário; o email sai sempre em português por agora
 *   (a notificação in-app já era traduzida do lado do cliente, essa parte
 *   não muda).</li>
 *   <li>Eventos de domínios ainda não portados (subscrição, ERP,
 *   Supplier Development, nota de crédito, apólices, cadastro de
 *   empresa) — os métodos existem no Node mas não têm chamador em Java
 *   enquanto esses domínios não existirem; não são replicados aqui até
 *   terem quem os invoque.</li>
 * </ul>
 */
@Service
public class NotificationService {

    private static final DateTimeFormatter DATA = DateTimeFormatter.ofPattern("yyyy-MM-dd").withZone(ZoneOffset.UTC);

    private final NotificationRepository notificationRepository;
    private final UserRepository userRepository;
    private final CompanyRepository companyRepository;
    private final EmailDispatchService emailDispatchService;

    public NotificationService(NotificationRepository notificationRepository, UserRepository userRepository,
                                CompanyRepository companyRepository, EmailDispatchService emailDispatchService) {
        this.notificationRepository = notificationRepository;
        this.userRepository = userRepository;
        this.companyRepository = companyRepository;
        this.emailDispatchService = emailDispatchService;
    }

    public Notification notifyUser(String userId, NotificationType type, String title, String message,
                                    NotificationChannel channel, String relatedEntityType, String relatedEntityId,
                                    String emailTo) {
        Notification notification = new Notification(UUID.randomUUID().toString(), userId, null, type, channel,
                title, message, relatedEntityType, relatedEntityId, Instant.now());
        notificationRepository.save(notification);

        if (channel == NotificationChannel.IN_APP_EMAIL && emailTo != null && !emailTo.isBlank()) {
            // O EMAIL sai sempre em português por agora — ver Javadoc da classe (i18n/emails.js não portado).
            emailDispatchService.dispatch(emailTo, title, message);
        }

        // TODO (M6): realtimeService.emitToUser(userId, "notification:new", notification) — STOMP.
        return notification;
    }

    public List<Notification> notifyUsersByRole(String companyId, List<PersonaRole> roles, NotificationType type,
                                                  String title, String message, NotificationChannel channel,
                                                  String relatedEntityType, String relatedEntityId) {
        List<User> users = userRepository.findByCompanyIdAndRoleInAndActiveTrue(companyId, roles);
        return users.stream()
                .map(u -> notifyUser(u.getId(), type, title, message, channel, relatedEntityType, relatedEntityId, u.getEmail()))
                .toList();
    }

    public Notification notifyCompanyContact(String companyId, NotificationType type, String title, String message) {
        Company company = companyRepository.findById(companyId).orElse(null);
        if (company == null) return null;

        Notification notification = new Notification(UUID.randomUUID().toString(), null, companyId, type,
                NotificationChannel.EMAIL, title, message, null, null, Instant.now());
        notificationRepository.save(notification);
        emailDispatchService.dispatch(company.getContactEmail(), title, message);
        return notification;
    }

    // --- Eventos de negócio (mapeados 1:1 com a secção 6 da especificação) -----

    public void poAguardaAprovacao(PurchaseOrder po) {
        notifyUsersByRole(po.getBuyerCompanyId(), List.of(PersonaRole.COMPANY_ADMIN), NotificationType.PO_AGUARDA_APROVACAO,
                "Nova PO aguarda aprovação", "A ordem de compra " + po.getReference() + " aguarda a sua aprovação.",
                NotificationChannel.IN_APP, "PurchaseOrder", po.getId());
    }

    public void poAprovadaOuRejeitada(PurchaseOrder po) {
        boolean aprovada = po.getStatus() == PoStatus.APROVADA;
        notifyUser(po.getCreatedById(), aprovada ? NotificationType.PO_APROVADA : NotificationType.PO_REJEITADA,
                aprovada ? "PO aprovada" : "PO rejeitada",
                "A ordem de compra " + po.getReference() + " foi " + (aprovada ? "aprovada" : "rejeitada") + ".",
                NotificationChannel.IN_APP, "PurchaseOrder", po.getId(), null);
    }

    public void poRecebidaPeloFornecedor(PurchaseOrder po) {
        notifyUsersByRole(po.getSupplierCompanyId(), List.of(PersonaRole.FORNECEDOR, PersonaRole.COMPANY_ADMIN),
                NotificationType.PO_RECEBIDA_FORNECEDOR, "Nova ordem de compra recebida",
                "Recebeu a ordem de compra " + po.getReference() + ". Reveja e aceite ou recuse.",
                NotificationChannel.IN_APP_EMAIL, "PurchaseOrder", po.getId());
    }

    public void poRecusadaPeloFornecedor(PurchaseOrder po) {
        notifyUsersByRole(po.getBuyerCompanyId(), List.of(PersonaRole.COMPRADOR, PersonaRole.COMPANY_ADMIN),
                NotificationType.PO_RECUSADA_FORNECEDOR, "PO recusada pelo fornecedor",
                "A ordem de compra " + po.getReference() + " foi recusada pelo fornecedor. Motivo: " + po.getRefusalReason(),
                NotificationChannel.IN_APP_EMAIL, "PurchaseOrder", po.getId());
    }

    public void faturaGerada(Invoice invoice, PurchaseOrder po) {
        String prazo = po.getPaymentDueAt() == null ? "" : DATA.format(po.getPaymentDueAt());
        notifyUsersByRole(po.getBuyerCompanyId(), List.of(PersonaRole.FINANCEIRO), NotificationType.FATURA_GERADA,
                "Fatura pendente de pagamento",
                "A fatura " + invoice.getReference() + " (PO " + po.getReference() + ") foi gerada. Prazo de pagamento: " + prazo + ".",
                NotificationChannel.IN_APP_EMAIL, "Invoice", invoice.getId());
    }

    public void entregaDespachada(PurchaseOrder po) {
        notifyUsersByRole(po.getBuyerCompanyId(), List.of(PersonaRole.COMPRADOR), NotificationType.ENTREGA_DESPACHADA,
                "Entrega despachada", "A entrega da PO " + po.getReference() + " foi despachada.",
                NotificationChannel.IN_APP, "PurchaseOrder", po.getId());
    }

    public void poEntregue(PurchaseOrder po) {
        notifyUsersByRole(po.getBuyerCompanyId(), List.of(PersonaRole.COMPRADOR, PersonaRole.COMPANY_ADMIN),
                NotificationType.PO_ENTREGUE, "Entrega marcada como concluída pelo fornecedor",
                "O fornecedor marcou a entrega da PO " + po.getReference() + " como concluída. Confirme a receção.",
                NotificationChannel.IN_APP_EMAIL, "PurchaseOrder", po.getId());
    }

    public void poRecebidaConforme(PurchaseOrder po) {
        notifyUsersByRole(po.getSupplierCompanyId(), List.of(PersonaRole.FORNECEDOR, PersonaRole.COMPANY_ADMIN),
                NotificationType.PO_RECEBIDA_CONFORME, "Receção confirmada sem divergências",
                "O comprador confirmou a receção da PO " + po.getReference() + " sem divergências. A ordem foi concluída.",
                NotificationChannel.IN_APP_EMAIL, "PurchaseOrder", po.getId());
    }

    public void rececaoComDivergencia(PurchaseOrder po) {
        notifyUsersByRole(null, List.of(PersonaRole.ADMIN_SISTEMA), NotificationType.RECECAO_COM_DIVERGENCIA,
                "Receção com divergência reportada",
                "A PO " + po.getReference() + " foi recebida com divergência (\"" + po.getReceptionStatus()
                        + "\"). Caso a acompanhar fora da plataforma (sinistro).",
                NotificationChannel.IN_APP_EMAIL, "PurchaseOrder", po.getId());
    }

    public void divergenciaResolvida(PurchaseOrder po) {
        boolean reposicao = "REPOSICAO".equals(po.getDivergenceResolution());
        String notas = po.getDivergenceResolutionNotes();
        notifyUsersByRole(po.getSupplierCompanyId(), List.of(PersonaRole.FORNECEDOR, PersonaRole.COMPANY_ADMIN),
                NotificationType.DIVERGENCIA_RESOLVIDA, reposicao ? "Reposição solicitada" : "Divergência resolvida",
                reposicao
                        ? "A divergência da PO " + po.getReference() + " foi resolvida com pedido de REPOSIÇÃO: "
                        + "corrija/reentregue a mercadoria e volte a marcar como entregue."
                        + (notas != null && !notas.isBlank() ? " Notas: " + notas : "")
                        : "A divergência da PO " + po.getReference() + " foi resolvida: o comprador aceitou a entrega e a ordem foi concluída.",
                NotificationChannel.IN_APP_EMAIL, "PurchaseOrder", po.getId());

        notifyUsersByRole(null, List.of(PersonaRole.ADMIN_SISTEMA), NotificationType.DIVERGENCIA_RESOLVIDA,
                "Divergência resolvida",
                "A divergência da PO " + po.getReference() + " foi resolvida (" + (reposicao ? "reposição" : "aceite") + ").",
                NotificationChannel.IN_APP, "PurchaseOrder", po.getId());
    }

    public void estoqueBaixo(String supplierId, String productId, String productName, Integer stockQuantity, Integer minStock) {
        notifyUsersByRole(supplierId, List.of(PersonaRole.FORNECEDOR, PersonaRole.COMPANY_ADMIN), NotificationType.ESTOQUE_BAIXO,
                "Stock abaixo do mínimo",
                "O produto \"" + productName + "\" está com " + (stockQuantity == null ? 0 : stockQuantity)
                        + " unidades em stock, abaixo do mínimo definido (" + minStock + "). Considere repor.",
                NotificationChannel.IN_APP, "Product", productId);
    }

    public void supplierDevRecebida(SupplierDevRequest r) {
        notifyUsersByRole(null, List.of(PersonaRole.ADMIN_SISTEMA), NotificationType.SUPPLIER_DEV_RECEBIDA,
                "Nova candidatura ao Supplier Development",
                r.getCompanyName() + " candidatou-se ao programa (" + r.getReference() + "). Percurso: " + r.getTrack()
                        + ". Contacto: " + r.getContactName() + " — " + r.getContactEmail() + ". Taxa de acesso de "
                        + r.getAccessFeeUsd() + " USD emitida na submissão — por receber.",
                NotificationChannel.IN_APP_EMAIL, "SupplierDevRequest", r.getId());
    }
}
